### PR TITLE
[pkg/ottl] Update NewParser to handle nil logger

### PR DESCRIPTION
**Pull request by TylerHelmuth**
**Description:** 
Updates the NewParser function to an error if the `component.TelemetrySettings` struct that is passed in does not have a logger.  This prevents https://github.com/open-telemetry/opentelemetry-collector-contrib/blob/eedb37161a6420f3670897e347c75b8f66637b7d/pkg/ottl/compare.go#LL32C2-L32C2 from panicking.

**Testing:**
Ran tests

### CODE REVIEW

**Comment by atoulme** [comment]
I think from experience that the feedback you might get here is to make sure you don't mutate the TelemetrySettings object, and rather create a new one instead by copying elements into it.

**Review by djaglowski — APPROVED** [review]

**Comment by TylerHelmuth** [comment]
> I think from experience that the feedback you might get here is to make sure you don't mutate the TelemetrySettings object, and rather create a new one instead by copying elements into it.

@atoulme isn't the function mutating a copy of the original struct?  Changing the struct within NewParser should not change the struct that was passed into the function.  If it is a bad practice anyways I can update it to make a new struct if a logger needs to be provided.

**Comment by atoulme** [comment]
No strong opinion, and this can stand for now I think, but do take a look at this issue for the context I am referring to: https://github.com/open-telemetry/opentelemetry-collector-contrib/issues/18044

**Comment by TylerHelmuth** [comment]
@atoulme I believe the difference with that issue is that `config` is a `*Config` so modifying `config.HTTPClientSettings.Compression = ""` does change the Compression value of the struct that was passed in.

**Review by evan-bradley — APPROVED** [review]

File: pkg/ottl/parser.go [hunk 1]
```diff
@@ -69,6 +69,14 @@ func NewParser[K any](
 	settings component.TelemetrySettings,
 	options ...Option[K],
 ) Parser[K] {
+	if settings.Logger == nil {
```
Comment (codeboten) [inline pkg/ottl/parser.go:72]: This shouldn't happen outside of tests correct? I'm curious about the choice to set a logger here. The other instances where the logger isn't set usually return an error condition.

Comment (TylerHelmuth) [reply pkg/ottl/parser.go:72]: Mostly likely only testing, although the package is exported publicly and can be used everywhere. 

If it is more idiomatic I can update the function to return an error.

Comment (codeboten) [reply pkg/ottl/parser.go:72]: Based on the rest of the codebase I would lean towards returning an error here.

Comment (TylerHelmuth) [reply pkg/ottl/parser.go:72]: Switched to return an error

**Review by djaglowski — APPROVED** [review]

**Pull request merged**